src/reader/decompress.rs

```diff
@@ -21,6 +70,53 @@ fn decode_header(bytes: &[u8]) -> (bool, usize) {
     (is_original, length)
 }
 
+fn decompress_block(
+    compression: Compression,
+    compressed_bytes: &[u8],
+    scratch: &mut Vec<u8>,
+) -> Result<(), Error> {
+    match compression.compression_type {
+        CompressionType::Zlib => {
+            let mut gz = flate2::read::DeflateDecoder::new(compressed_bytes);
+            scratch.clear();
+            gz.read_to_end(scratch).context(error::IoSnafu)?;
+        }
+        CompressionType::Zstd => {
+            let mut reader =
+                zstd::Decoder::new(compressed_bytes).context(error::BuildZstdDecoderSnafu)?;
+            scratch.clear();
+            reader.read_to_end(scratch).context(error::IoSnafu)?;
+        }
+        CompressionType::Snappy => {
+            let len = snap::raw::decompress_len(compressed_bytes)
+                .context(error::BuildSnappyDecoderSnafu)?;
+            scratch.resize(len, 0);
+            let mut decoder = snap::raw::Decoder::new();
+            decoder
+                .decompress(compressed_bytes, scratch)
+                .context(error::BuildSnappyDecoderSnafu)?;
+        }
+        CompressionType::Lzo => {
+            let decompressed = lzokay_native::decompress_all(compressed_bytes, None)
+                .context(error::BuildLzoDecoderSnafu)?;
+            // TODO: better way to utilize scratch here
+            scratch.clear();
+            scratch.extend(decompressed);
+        }
+        CompressionType::Lz4 => {
+            let decompressed = lz4_flex::block::decompress(
+                compressed_bytes,
+                compression.max_decompressed_block_size,
+            )
+            .context(error::BuildLz4DecoderSnafu)?;
+            // TODO: better way to utilize scratch here
+            scratch.clear();
+            scratch.extend(decompressed);
+        }
```

Comment:
There's probably ways to make this more efficient but can probably tackle that later, especially when we have benchmarks to be able to measure such improvements